mri_quality_check now returns list of errors which are added to empty…

gcyoung1 opened the pull request:
… scan notes for unknown quality scans

gcyoung1 commented:
Tests To Do:
- [ ] Non-empty scan notes are left as is
- [ ] Questionable scans left as-is
- [ ] Usable scans left as-is
- [ ] Unusable scans left as-is
- [ ] Scan with multiple issues
- [ ] Formatting on xnat

kipohl commented:
@gcyoung1 most changes seem stylistic - can you point me to the ones I have to review - also can you print out an output when you run it ?

gcyoung1 commented:
@kipohl The first commit has all the substantive changes. I'm not sure exactly how to test it because there are only 2 currently existing scan errors in production and when I run check_new_sessions on them, they don't show up in the unseen_scans list.

kipohl commented:
@gcyoung1 set their labels to unknown - run the test and then return it to whatever status it was before . Does that make sense ? Also can you point me to the lines that are not stylistic changes ?  Also why did you make all those stylistic changes ?

gcyoung1 commented:
@kipohl I'm not sure how to get scans with errors.
The second commit was just me linting--no substantive changes in that one.

gcyoung1 commented:
Testing on [sibis](https://sibis.sri.com/xnattest/app/template/XDATScreen_report_xnat_mrSessionData.vm/search_field/xnat%3AmrSessionData.id/search_value/NCANDA_E03249/search_element/xnat%3AmrSessionData/popup/false/project/upmc_incoming):
I set scans 2, 12, and 13 to questionable. I will add one error to scan 2 and scan 12 and 2 errors to scan 13. I added a scan note to scan 2, so its scan note should not be updated.
I set scan 3 to questionable, 4 to usable,and 6 to unusable.

ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_not
es)$                  python -m pdb check_new_sessions  -e NCANDA_E03249
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(8)<module>()
-> from __future__ import print_function
(Pdb) b 684
Breakpoint 1 at /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions:684
(Pdb) c
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(684)mri_quality_check()
-> return errors
(Pdb) error_12 =  {
                    "scan_id": 12, 
                    "scan_type": 'scan_type_test',
                    "expected_nii_files": '1',
                    "num_nii_files": '2',
                    "manufacturer": 'manufacturer_test',
                    "nii_search": 'nii_search_test',
                } 
(Pdb) error_13_1 =  {
                    "scan_id": 13, 
                    "scan_type": 'scan_type_test',
                    "expected_nii_files": '1',
                    "num_nii_files": '2',
                    "manufacturer": 'manufacturer_test',
                    "nii_search": 'nii_search_test',
                } 
(Pdb) error_13_2 =  {
                    "scan_id": 13, 
                    "scan_type": 'scan_type_test',
                    "expected_nii_files": '1',
                    "num_nii_files": '2',
                    "manufacturer": 'manufacturer_test',
                    "nii_search": 'nii_search_test',
                } 
(Pdb) errors = [error_12, error_13_1, error_13_2]
(Pdb) b 1513
Breakpoint 2 at /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions:1513
(Pdb) c
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1513)<module>()
-> scan_note = sanitize_string_for_csv(scan_note)
(Pdb) scan
'2'
(Pdb) scan_note
'I&apos;m not empty'
(Pdb) c
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1513)<module>()
-> scan_note = sanitize_string_for_csv(scan_note)
(Pdb) scan
'12'
(Pdb) scan_note
"Scan Errors:\n[{'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}]"
(Pdb) c
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1513)<module>()
-> scan_note = sanitize_string_for_csv(scan_note)
(Pdb) scan
'13'
(Pdb) scan_note
"Scan Errors:\n[{'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}, {'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}]"



In conclusion: 
- [ ] Non-empty scan notes are left as is
- [ ] Questionable scans left as-is
- [ ] Usable scans left as-is
- [ ] Unusable scans left as-is
- [ ] Scan with multiple issues
- [ ] Formatting on xnat

kipohl commented:
@gcyoung1 what is 'nii_search': 'nii_search_test' ? 

Also if the manufacturer is incorrect then the program does not behave correctly -

```
if not manufacturer in list(expected_images.keys()):
...
  return []
(line 494)

mri_errors = mri_quality_check(experiment, session_dir, manufacturer, session_label, eid, scan_id_list, incorrect_formatting_map)
errorFlag = errorFlag or (len(mri_errors) != 0) 
```
Finally, let me know when you take are of the xnat formatting so I can see how the string would be printed out - in general the short the message the better as the summary view in xnat only displays a view characters

kipohl commented:
also run please 
```
check_new_sessions -e NCANDA_E03249
```
so I can easily see what would be spit out

gcyoung1 commented:
All the errors in the test were created by stepping into the check_mri_quality function and setting the list of errors to be some made up ones so I could see how the scan notes would be created. So 'nii_search': 'nii_search_test' is just a dummy field I made.

Should I add the manufacturer error to the experiment notes? I'm assuming that I shouldn't in this next commit.

I'm not sure how the csv files are uploaded to xnat. Currently they're just discarded unless you use the --qc-csv flag, in which case they're written to the check_new_sessions log_dir, but I'm not sure how to upload them to xnat from there.

gcyoung1 commented:
This is the output on sibis:
ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ ./check_new_sessions -e NCANDA_E03249 -v

Script was last run 2021-02-25 17:11:38.507
Re-checking 1 previously flagged experiments
Checking sessions modified after 2021-02-25 17:11:38
Checking a total of 1 experiments
1 Checking experiment upmc_incoming/A-00002-F-2-20150303 SID:NCANDA_S00061 EID:NCANDA_E03249, insert date 2015-03-11 08:53:38.079, modified date 2022-04-11 21:33:09.584501...
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 1 ncanda-localizer-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 2 ncanda-t2fse-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 3 ncanda-mprage-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 4 ncanda-dti6b500pepolar-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 6 ncanda-dti60b1000-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 12 ncanda-grefieldmap-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 13 ncanda-grefieldmap-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 14 ncanda-rsfmri-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 15 ncanda-fmri-ringrewards-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 16 ncanda-fmri-ringrewards-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 18 ncanda-fmri-ringrewards-v2 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 20 ncanda-fmri-ringrewards-v3 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 22 ncanda-fmri-ringrewards-v4 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Beginning QC...{"experiment_site_id": "A-00002-F-2-20150303-9286b4", "error": "[Errno 2] No such file or directory: '/fs/ncanda-share/log/check_new_sessions/passed_qc/NCANDA_E03249_passed_qc'", "project": "upmc_incoming", "file": "/fs/ncanda-share/log/check_new_sessions/passed_qc/NCANDA_E03249_passed_qc", "eid": "NCANDA_E03249"}
RECHECK
<b>Summary</b>
<ol>
<li>Sessions with Scans to Inspect: <a href="#toinspect">&nbsp;1&nbsp;</a></li>
<li>Sessions with Questionable Scans: <a href="#questionable">&nbsp;1&nbsp;</a></li>
</ol>
<b><a id="toinspect">Sessions with Mandatory Scans to Inspect</a></b>
<ol><li><a href=https://sibis.sri.com/xnattest/app/action/DisplayItemAction/search_value/NCANDA_E03249/search_element/xnat:mrSessionData/search_field/xnat:mrSessionData.ID/project/upmc_incoming>upmc_incoming/NCANDA_E03249/A-00002-F-2-20150303</a></li>
/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti<BR>
ncanda-t2fse-v1, ncanda-grefieldmap-v1, ncanda-grefieldmap-v1</ol><b><a id="questionable">Sessions with Questionable Scans</a></b>
<ol><li><a href=https://sibis.sri.com/xnattest/app/action/DisplayItemAction/search_value/NCANDA_E03249/search_element/xnat:mrSessionData/search_field/xnat:mrSessionData.ID/project/upmc_incoming>upmc_incoming/NCANDA_E03249/A-00002-F-2-20150303</a></li>
/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti<BR>
ncanda-mprage-v1</ol>







However, when I tested it on ncanda, I got the following error:
ncanda@pipeline-back[pipeline_back_1]:/sibis-software/ncanda-data-integration/scripts/xnat (master)$ /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions -e NCANDA_E03249 -v
Script was last run 2022-04-13 12:34:58.381
Re-checking 1 previously flagged experiments
Checking sessions modified after 2022-04-13 12:34:58
Checking a total of 1 experiments
1 Checking experiment upmc_incoming/A-00002-F-2-20150303 SID:NCANDA_S00061 EID:NCANDA_E03249, insert date 2015-03-11 08:53:38.079, modified date 2015-04-07 14:23:06.850767...
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 1 ncanda-localizer-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 2 ncanda-t2fse-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 3 ncanda-mprage-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 4 ncanda-dti6b500pepolar-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 6 ncanda-dti60b1000-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 12 ncanda-grefieldmap-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 13 ncanda-grefieldmap-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 14 ncanda-rsfmri-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 15 ncanda-fmri-ringrewards-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 16 ncanda-fmri-ringrewards-v1 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 18 ncanda-fmri-ringrewards-v2 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 20 ncanda-fmri-ringrewards-v3 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Starting export of nifti files for  NCANDA_S00061 NCANDA_E03249 A-00002-F-2-20150303 22 ncanda-fmri-ringrewards-v4 /fs/storage/XNAT
... nothing to do as nifti files are up to date
Beginning QC...Skipping QC...Traceback (most recent call last):
  File "/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions", line 1502, in <module>
    for error in scan_errors:
NameError: name 'scan_errors' is not defined





I believe this is because qc_file_tmp exists already so the mri_quality_check is not done. I just added a statement to initialize the scan errors to an empty list so it's always defined.

kipohl commented:
regarding csv: they are later uploaded by check_new_sessions the next time it runs on all cases - no worries 
regarding manufacturing error : 
1) if only the manufacture is wrong the function mri_quality_check would return an empty error log which implies that function did not detect an error - which is wrong 
2)  Yes ideally those would be added to experimental notes if they are empty but this would be harder bc you would have to write it directly to xnat and not via the csv file so if it is not easy for you simply do not write it to xnat and only raise the github issue

kipohl commented:
@gcyoung1 forgot to mention you

kipohl commented:
@gcyoung1 I see that your newest commit now handle the issue correctly. Could you run the check_new_sessions  (without the -p option) and print the output (omitting pdb) as well as the csv and csv2 file ?

kipohl commented:
@gcyoung1 : do not understand why it work on testing but not production 
also on your test enviornment can you make sure that '/fs/ncanda-share/log/check_new_sessions/passed_qc/' exists with the right access rights so you do not get an error at the end ?

kipohl commented:
can you run it one more time without the -v option now

kipohl commented:
@gcyoung1 so when you run check_new_sessions on an indvidual experiment please make sure you run with -f option otherwise it might not be checked- this is especially important when you remove items from special_cases.yml

kipohl commented:
as you see in your prior run in the output it states "Beginning QC...Skipping QC..." - so it never checked it

gcyoung1 commented:
@kipohl '/fs/ncanda-share/log/check_new_sessions/passed_qc/' does not exist on sibis:

ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ ./check_new_sessions -e NCANDA_E03249 --qc-csv --qc-csv-v2 -f
{"experiment_site_id": "A-00002-F-2-20150303-9286b4", "error": "[Errno 2] No such file or directory: '/fs/ncanda-share/log/check_new_sessions/passed_qc/NCANDA_E03249_passed_qc'", "project": "upmc_incoming", "file": "/fs/ncanda-share/log/check_new_sessions/passed_qc/NCANDA_E03249_passed_qc", "eid": "NCANDA_E03249"}


Also, I need to be in pdb to introduct fake errors to be added to the scan notes.

kipohl commented:
That is why I said that you should create it

On Thu, Apr 14, 2022, 5:08 PM gcyoung1 ***@***.***> wrote:

> @kipohl <https://github.com/kipohl>
> '/fs/ncanda-share/log/check_new_sessions/passed_qc/' does not exist on
> sibis:
>
> ***@***.***[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat
> (feature/add_issues_to_scan_notes)$ ./check_new_sessions -e NCANDA_E03249
> --qc-csv --qc-csv-v2 -f
> {"experiment_site_id": "A-00002-F-2-20150303-9286b4", "error": "[Errno 2]
> No such file or directory:
> '/fs/ncanda-share/log/check_new_sessions/passed_qc/NCANDA_E03249_passed_qc'",
> "project": "upmc_incoming", "file":
> "/fs/ncanda-share/log/check_new_sessions/passed_qc/NCANDA_E03249_passed_qc",
> "eid": "NCANDA_E03249"}
>
> Also, I need to be in pdb to introduct fake errors to be added to the scan
> notes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sibis-platform/ncanda-data-integration/pull/468#issuecomment-1099808063>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRYGEAUFAXLU72YNPGLC7TVFDMSFANCNFSM5S5WL46Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gcyoung1 commented:
Ok, I created the directory.
Here's the output when I don't introduce any erros with pdb:

ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ ./check_new_sessions -e NCANDA_E03249 --qc-csv --qc-csv-v2 -f
ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ cat /fs/ncanda-share/log/check_new_sessions/scans_to_review.csv
xnat_experiment_id,nifti_folder,scan_id,scan_type,experiment_note,decision,scan_note
NCANDA_E03249,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti,2,ncanda-t2fse-v1,"",,"I&amp;amp;amp;amp;amp;amp;amp;apos;m not empty"
NCANDA_E03249,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti,12,ncanda-grefieldmap-v1,"",,""
NCANDA_E03249,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti,13,ncanda-grefieldmap-v1,"",,""
ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ cat /fs/ncanda-share/log/check_new_sessions/scans_to_review_
scans_to_review_miqa4.csv  scans_to_review_v2.csv
ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ cat /fs/ncanda-share/log/check_new_sessions/scans_to_review_v2.csv 
project_name,experiment_name,scan_name,scan_type,frame_number,file_location
upmc,NCANDA_E03249,NCANDA_E03249_ncanda-t2fse-v1,ncanda-t2fse-v1,2,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti/2_ncanda-t2fse-v1/image.nii.gz
upmc,NCANDA_E03249,NCANDA_E03249_ncanda-grefieldmap-v1,ncanda-grefieldmap-v1,12,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti/12_ncanda-grefieldmap-v1/image.nii.gz
upmc,NCANDA_E03249,NCANDA_E03249_ncanda-grefieldmap-v1,ncanda-grefieldmap-v1,13,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti/13_ncanda-grefieldmap-v1/image.nii.gz

gcyoung1 commented:
And here's the output with intoduced errors. Note that scan 2, which already had a scan note, is left untouched:

ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ python -m pdb check_new_sessions -e NCANDA_E03249 --qc-csv --qc-csv-v2 -f
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(8)<module>()
-> from __future__ import print_function
(Pdb) b 669
Breakpoint 1 at /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions:669
(Pdb) c
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(669)mri_quality_check()
-> if scan_errors:
(Pdb) error_12 = {
"scan_id": 12,
"scan_type": 'scan_type_test',
"expected_nii_files": '1',
"num_nii_files": '2',
"manufacturer": 'manufacturer_test',
"nii_search": 'nii_search_test',
}
(Pdb) error_2 = {
"scan_id": 2,
"scan_type": 'scan_type_test',
"expected_nii_files": '1',
"num_nii_files": '2',
"manufacturer": 'manufacturer_test',
"nii_search": 'nii_search_test',
}
(Pdb) scan_errors = [error_2, error_12]
(Pdb) c
{"experiment_site_id": "A-00002-F-2-20150303", "error": "ERROR: scans in session failed to match standard setting", "violations": "[{'scan_id': 2, 'scan_type': 'scan_type_test', 'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}, {'scan_id': 12, 'scan_type': 'scan_type_test', 'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}]", "eid": "NCANDA_E03249", "site_forward": "upmc", "site_resolution": "If the scan is expected to be wrong, please let the Datacore know so that they can set an exception. If the scan wasn't expected to be wrong and you believe this error may be due to an upload failure, please re-upload. If the re-uploaded scan continues to exhibit the problematic features, please inquire with the Datacore to check whether a re-scan should be scheduled."}
The program finished and will be restarted
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(8)<module>()
-> from __future__ import print_function
(Pdb) q
ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-
data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ cat /fs/ncanda-share/log/check_new_sessions/scans_to_review.csv 
xnat_experiment_id,nifti_folder,scan_id,scan_type,experiment_note,decision,scan_note
NCANDA_E03249,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti,2,ncanda-t2fse-v1,"Error: Manufacturer unknown:&apos;test_manufacturer&apos;",,"I&amp;amp;amp;amp;amp;amp;amp;amp;amp;amp;apos;m not empty"
NCANDA_E03249,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti,12,ncanda-grefieldmap-v1,"Error: Manufacturer unknown:&apos;test_manufacturer&apos;",,"Scan Errors: [{'expected_nii_files': '1' - 'num_nii_files': '2' - 'manufacturer': 'manufacturer_test' - 'nii_search': 'nii_search_test'}]"
NCANDA_E03249,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti,13,ncanda-grefieldmap-v1,"Error: Manufacturer unknown:&apos;test_manufacturer&apos;",,""
ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ cat /fs/ncanda-share/log/check_new_sessions/scans_to_review_
scans_to_review_miqa4.csv  scans_to_review_v2.csv
ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ cat /fs/ncanda-share/log/check_new_sessions/scans_to_review_v2.csv 
project_name,experiment_name,scan_name,scan_type,frame_number,file_location
upmc,NCANDA_E03249,NCANDA_E03249_ncanda-t2fse-v1,ncanda-t2fse-v1,2,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti/2_ncanda-t2fse-v1/image.nii.gz
upmc,NCANDA_E03249,NCANDA_E03249_ncanda-grefieldmap-v1,ncanda-grefieldmap-v1,12,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti/12_ncanda-grefieldmap-v1/image.nii.gz
upmc,NCANDA_E03249,NCANDA_E03249_ncanda-grefieldmap-v1,ncanda-grefieldmap-v1,13,/fs/storage/XNAT/archive/upmc_incoming/arc001/A-00002-F-2-20150303/RESOURCES/nifti/13_ncanda-grefieldmap-v1/image.nii.gz

kipohl commented:
can you actually attach the two csv files to the pull request ?

kipohl commented:
@gcyoung1 also when you post these messages can you please use 
```

```
so they are formatted better

kipohl commented:
I did it now for the last entry so that I can read the output more easily - why does the scan notes have so many special characters - that might produce problems 
```
Manufacturer unknown:&apos;test_manufacturer&apos;",,"I&amp;amp;amp;amp;amp;amp;amp;amp;amp;amp;apos;m not empty
```

kipohl commented:
we should not have any special characters in those notes as this can produce problems when transfering notes to redcap

gcyoung1 commented:
@kipohl Yes, xnat does not seem to handle the special characters well. I will remove them from the error messages that are posted.

gcyoung1 commented:
```
ncanda@pipeline-back[pipeline_back_1]:/fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat (feature/add_issues_to_scan_notes)$ python -m pdb ./check_new_sessions -e NCANDA_E03249 --qc-csv --qc-csv-v2
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(8)<module>()
-> from __future__ import print_function
(Pdb) b 669
Breakpoint 1 at /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions:669
(Pdb) c
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(669)mri_quality_check()
-> if scan_errors:
(Pdb)  error_12 = {
"scan_id": 12,
"scan_type": 'scan_type_test',
"expected_nii_files": '1',
"num_nii_files": '2',
"manufacturer": 'manufacturer_test',
"nii_search": 'nii_search_test',
}
(Pdb)  error_2_1 = {
"scan_id": 2,
"scan_type": 'scan_type_test',
"expected_nii_files": '1',
"num_nii_files": '2',
"manufacturer": 'manufacturer_test',
"nii_search": 'nii_search_test',
}
(Pdb)  error_2_2 = {
"scan_id": 2,
"scan_type": 'scan_type_test',
"expected_nii_files": '1',
"num_nii_files": '2',
"manufacturer": 'manufacturer_test',
"nii_search": 'nii_search_test',
}
(Pdb) scan_errors = [error_12, error_2_1, error_2_2]
(Pdb) b 1513
Breakpoint 2 at /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions:1513
(Pdb) c
{"experiment_site_id": "A-00002-F-2-20150303", "error": "ERROR: scans in session failed to match standard setting", "violations": "[{'scan_id': 12, 'scan_type': 'scan_type_test', 'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}, {'scan_id': 2, 'scan_type': 'scan_type_test', 'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}, {'scan_id': 2, 'scan_type': 'scan_type_test', 'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}]", "eid": "NCANDA_E03249", "site_forward": "upmc", "site_resolution": "If the scan is expected to be wrong, please let the Datacore know so that they can set an exception. If the scan wasn't expected to be wrong and you believe this error may be due to an upload failure, please re-upload. If the re-uploaded scan continues to exhibit the problematic features, please inquire with the Datacore to check whether a re-scan should be scheduled."}
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1513)<module>()
-> smaller_error = exclude_keys(error, keys_to_exclude)
(Pdb) error
{'scan_id': 12, 'scan_type': 'scan_type_test', 'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}
(Pdb) n
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1514)<module>()
-> error_string = dict_to_string(smaller_error)
(Pdb) errors_by_scan['12']
''
(Pdb) n
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1515)<module>()
-> scan_id = str(error["scan_id"])
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1516)<module>()
-> errors_by_scan[scan_id] = errors_by_scan[scan_id] + error_string
(Pdb) error_string
'expected_nii_files:\n1\nnum_nii_files:\n2\nnii_search:\nnii_search_test\n'
(Pdb) scan_id
'12'
(Pdb) scan_errors
[{'scan_id': 12, 'scan_type': 'scan_type_test', 'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}, {'scan_id': 2, 'scan_type': 'scan_type_test', 'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}, {'scan_id': 2, 'scan_type': 'scan_type_test', 'expected_nii_files': '1', 'num_nii_files': '2', 'manufacturer': 'manufacturer_test', 'nii_search': 'nii_search_test'}]
(Pdb) n
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1512)<module>()
-> for error in scan_errors:
(Pdb) errors_by_scan['12']
'expected_nii_files:\n1\nnum_nii_files:\n2\nnii_search:\nnii_search_test\n'
(Pdb) n
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1513)<module>()
-> smaller_error = exclude_keys(error, keys_to_exclude)
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1514)<module>()
-> error_string = dict_to_string(smaller_error)
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1515)<module>()
-> scan_id = str(error["scan_id"])
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1516)<module>()
-> errors_by_scan[scan_id] = errors_by_scan[scan_id] + error_string
(Pdb) scan_id
'2'
(Pdb) n
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1512)<module>()
-> for error in scan_errors:
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1513)<module>()
-> smaller_error = exclude_keys(error, keys_to_exclude)
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1514)<module>()
-> error_string = dict_to_string(smaller_error)
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1515)<module>()
-> scan_id = str(error["scan_id"])
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1516)<module>()
-> errors_by_scan[scan_id] = errors_by_scan[scan_id] + error_string
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1512)<module>()
-> for error in scan_errors:
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1519)<module>()
-> if len(unseen_scans):
(Pdb) errors_by_scan
defaultdict(<class 'str'>, {'12': 'expected_nii_files:\n1\nnum_nii_files:\n2\nnii_search:\nnii_search_test\n', '2': 'expected_nii_files:\n1\nnum_nii_files:\n2\nnii_search:\nnii_search_test\nexpected_nii_files:\n1\nnum_nii_files:\n2\nnii_search:\nnii_search_test\n'})
(Pdb) b 1528
Breakpoint 3 at /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions:1528
(Pdb) c
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1528)<module>()
-> if error_string != "" and "Scan Errors:" not in scan_note:
(Pdb) scan
'2'
(Pdb) error_string
'expected_nii_files:\n1\nnum_nii_files:\n2\nnii_search:\nnii_search_test\nexpected_nii_files:\n1\nnum_nii_files:\n2\nnii_search:\nnii_search_test\n'
(Pdb) n
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1529)<module>()
-> scan_note += "\nScan Errors:\n" + error_string
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1530)<module>()
-> scan_note = sanitize_string_for_csv(scan_note)
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1532)<module>()
-> f'{eid},{session_dir},{scan},{scan_type},"{exp_note}",,"{scan_note}"\n'
(Pdb) scan_note
'This is a scan note Scan Errors: expected_nii_files: 1 num_nii_files: 2 nii_search: nii_search_test expected_nii_files: 1 num_nii_files: 2 nii_search: nii_search_test '
(Pdb) c
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1528)<module>()
-> if error_string != "" and "Scan Errors:" not in scan_note:
(Pdb) scan
'12'
(Pdb) error_string
'expected_nii_files:\n1\nnum_nii_files:\n2\nnii_search:\nnii_search_test\n'
(Pdb) scan_note
'Scan Errors:\r\nexpected_nii_files:\r\n1\r\nnum_nii_files:\r\n2\r\nnii_search:\r\nnii_search_test'
(Pdb) n
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1530)<module>()
-> scan_note = sanitize_string_for_csv(scan_note)
(Pdb) 
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1532)<module>()
-> f'{eid},{session_dir},{scan},{scan_type},"{exp_note}",,"{scan_note}"\n'
(Pdb) c
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1528)<module>()
-> if error_string != "" and "Scan Errors:" not in scan_note:
(Pdb) scan
'13'
(Pdb) error_string
''
(Pdb) n
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(1530)<module>()
-> scan_note = sanitize_string_for_csv(scan_note)
(Pdb) c
The program finished and will be restarted
> /fs/ncanda-share/beta/griffin/ncanda-data-integration/scripts/xnat/check_new_sessions(8)<module>()
-> from __future__ import print_function
(Pdb) q
Here are the csv files that were produced. As expected, scan 12 has no note added to it since it already had a scan error note. Scan 2 does have a note added to it because, although it has a note, it was not created by this script.
[xnat link](https://sibis.sri.com/xnattest/app/action/DisplayItemAction/search_element/xnat:mrSessionData/search_field/xnat:mrSessionData.ID/search_value/NCANDA_E03249/project/upmc_incoming)
[scans_to_review.csv](https://github.com/sibis-platform/ncanda-data-integration/files/8508636/scans_to_review.csv)
[scans_to_review_v2.csv](https://github.com/sibis-platform/ncanda-data-integration/files/8508637/scans_to_review_v2.csv)
@kipohl 

```

kipohl commented:
please do not include the following lines in formatted text so that it properly formatts them  

Here are the csv files that were produced. As expected, scan 12 has no note added to it since it already had a scan error note. Scan 2 does have a note added to it because, although it has a note, it was not created by this script.
[xnat link](https://sibis.sri.com/xnattest/app/action/DisplayItemAction/search_element/xnat:mrSessionData/search_field/xnat:mrSessionData.ID/search_value/NCANDA_E03249/project/upmc_incoming)
[scans_to_review.csv](https://github.com/sibis-platform/ncanda-data-integration/files/8508636/scans_to_review.csv)
[scans_to_review_v2.csv](https://github.com/sibis-platform/ncanda-data-integration/files/8508637/scans_to_review_v2.csv)

kipohl commented:
@gcyoung1 sorry am a little bit confused 
![image](https://user-images.githubusercontent.com/10715920/163893206-655cb6d9-9218-445e-8505-beb1faa242cc.png)
how do you know that the note was created by the script ? 
was the note 'Scan Errors: expected_nii_files: 1 num_nii_files: 2 nii_search: nii_search_test'

gcyoung1 commented:
Yes, that one was. When the script adds the scan errors to a note, it prepends the actual errors with the string "Scan Errors:". I can change the behavior so that the scan errors are only added if the note is completely empty if you want.

kipohl commented:
@gcyoung1 lets leave it right now that way 
just change 'Scan Errors' to 'ERR'  and add ';' at the end of the error message so it is clear where the message ends

kipohl commented:
can tell you more at 1 pm why I want it that way

gcyoung1 commented:
@kipohl done

kipohl commented:
@gcyoung1 please remove any \n line from notes - again no special characters 
"\nERR:\n"

gcyoung1 commented:
Replaced "\n" with "  " in both notes
